rust/chains/hyperlane-cosmos/src/providers/rpc.rs

```diff
@@ -84,84 +87,116 @@ impl WasmIndexer for CosmosWasmIndexer {
+            .and_lte("tx.height", *range.end() as u64)
+            .and_eq(
+                format!("{}-{}._contract_address", Self::WASM_TYPE, self.event_type),
+                contract_address.clone(),
+            );
+
+        println!("Query: {:?}", query.to_string());
+
+        let tx_search_result = client
+            .tx_search(query.clone(), false, 1, 30, Order::Ascending)
+            .await?;
+
+        let total_count = tx_search_result.total_count;
+        let last_page = total_count / 30 + (total_count % 30 != 0) as u32;
+
+        let handler = |txs: Vec<tx::Response>,
+                       block_hashs: HashMap<u64, H256>|
+         -> Vec<(T, LogMeta)> {
+            let mut result: Vec<(T, LogMeta)> = vec![];
+            let target_type = format!("{}-{}", Self::WASM_TYPE, self.event_type);
+
+            // Get BlockHash from block_search
+            let client = self.get_client().unwrap();
+
+            for tx in txs {
+                if tx.tx_result.code.is_err() {
+                    debug!("tx {:?} has failed. skip!", tx.hash);
                     continue;
                 }
 
-                if let Some(msg) = parser(event.attributes.clone()) {
-                    let meta = LogMeta {
-                        address: bech32_decode(addr.clone()),
-                        block_number: block_number as u64,
-                        block_hash: H256::from_slice(block.block_id.hash.as_bytes()),
-                        transaction_id: h256_to_h512(tx_hash),
-                        transaction_index: idx as u64,
-                        log_index: U256::from(log_idx),
-                    };
-
-                    parse_result.push((msg, meta));
+                let mut parse_result: Vec<(T, LogMeta)> = vec![];
+
+                for (log_idx, event) in tx.tx_result.events.clone().into_iter().enumerate() {
+                    if event.kind.as_str() == target_type {
+                        if let Some(msg) = parser(event.attributes.clone()) {
+                            let meta = LogMeta {
+                                address: bech32_decode(contract_address.clone()),
+                                block_number: tx.height.value(),
+                                block_hash: block_hashs[&tx.height.value()],
+                                transaction_id: h256_to_h512(H256::from_slice(tx.hash.as_bytes())),
+                                transaction_index: tx.index as u64,
+                                log_index: U256::from(log_idx),
+                            };
+
+                            parse_result.push((msg, meta));
+                        }
+                    }
                 }
-            }
 
-            if available {
                 result.extend(parse_result);
             }
+
+            result
+        };
+
+        let mut result = handler(tx_search_result.txs, block_hash.clone());
+
+        for page in 2..=last_page {
+            println!("page: {}", page);
```

Comment:
nit: just to avoid printlns in the agent, something like `debug!(page, "Making tx search RPC")` or something would be preferable